tests/test_api_typing.py

```diff
@@ -0,0 +1,220 @@
+"""Test module for classes in pandas.api.typing."""
+
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+from pandas._testing import ensure_clean
+from pandas.api.typing import (
+    DataFrameGroupBy,
+    DatetimeIndexResamplerGroupby,
+    Expanding,
+    ExpandingGroupby,
+    ExponentialMovingWindow,
+    ExponentialMovingWindowGroupby,
+    JsonReader,
+    NaTType,
+    NAType,
+    PeriodIndexResamplerGroupby,
+    Resampler,
+    Rolling,
+    RollingGroupby,
+    SeriesGroupBy,
+    TimedeltaIndexResamplerGroupby,
+    TimeGrouper,
+    Window,
+)
+import pytest
+from typing_extensions import assert_type
+
+from tests import check
+
+from pandas.io.json._json import read_json
+from pandas.io.stata import StataReader
+
+
+def test_dataframegroupby():
+    df = pd.DataFrame({"a": [1, 2, 3]})
+    group = df.groupby("a")
+
+    def f1(gb: DataFrameGroupBy):
+        check(gb, DataFrameGroupBy)
+
+    f1(group)
+
+
+def test_seriesgroupby():
+    sr = pd.Series([1, 2, 3], index=pd.Index(["a", "b", "a"]))
+
+    def f1(gb: SeriesGroupBy):
+        check(gb, SeriesGroupBy)
+
+    f1(sr.groupby(level=0))
+
+
+def tests_datetimeindexersamplergroupby() -> None:
+    idx = pd.date_range("1999-1-1", periods=365, freq="D")
+    df = pd.DataFrame(
+        np.random.standard_normal((365, 2)), index=idx, columns=["col1", "col2"]
+    )
+    gb_df = df.groupby("col2")
+
+    # TODO the groupby here is too wide and returns _ResamplerGroupby alias
+    # def f1(gb: DatetimeIndexResamplerGroupby):
+    #     check(gb, DatetimeIndexResamplerGroupby)
+    #
+    # f1(gb_df.resample("ME"))
+
+    check(gb_df.resample("ME"), DatetimeIndexResamplerGroupby)
```

Comment:
I was able to create a Union type for those three and add it to the function definition, though we can't use it as is in the check since it does not accept union types as klass but only individual type.